src/app.rs

```diff
@@ -65,13 +77,17 @@ impl App {
             }
         }
     }
-    /// Application main loop
+
+    /// Updates the UI and redraws the window (the applications main loop)
+    /// Event handling currently blocks the whole UI, eventually this should be
+    /// moved to asynchronous event handling.
```

Comment:
Using the word asynchronous is confusing here, because I think you just mean multi-threaded. There are good reasons why event handling is done on a single thread, I'll make a note to myself to update this doc comment with more info, for instance that it is necessary for event handlers to be responsible for not blocking the UI thread.

Reply:
I actually thought of using `futures` in order to enforce that the UI doesn't get blocked. The second reason is to not "buffer" events too much - this happens when the UI gets too many events and at some point you get a backlog of events that the actual application has to handle (doesn't happen if the application is fast enough). So my idea was: limn maintains a queue of all events (with configurable input masking, sort-of like X11) and an (ordered) list of futures that need to be resolved. In order for the application to handle an event, it has to take an event out of the queue and return a `future` to limn (which is a very quick operation and should not block in any significant way). Each frame / time interval, `limn` checks if there are futures that need to be resolved (or futures call `limn` if they are finished). If there are futures that have been resolved in this frame, the UI is redrawn. 

Of course, I have not thought about the details of this and how the state-tracking would work, but this was sort-of my plan. I'll remove "asynchronous" for now, but there should be some form of mechanism to ensure that the UI doesn't get blocked.

Reply:
Ok so would all this still be happening on one thread? How does limn know which futures need to be resolved and which don't? The nature of events is that they should all be intended to ultimately modify UI state, all of them may or may not trigger a redraw and it's not expected that there is a way to know beforehand (can you know when dispatching a key press whether it will eventually trigger a redraw?). Also there is no way to know whether an event handler will block (it's the halting problem) so the only way to ensure it doesn't block other stuff (the UI, rendering, input, whatever it is) is by putting it on a separate thread.

The issue of lag caused by a backlog of events being handled could be resolved by just checking if it's time to render after every event, and not care how many other events are waiting, but.. my feeling is that won't resolve the actual issue of event handlers blocking other events, so you might get a consistent frame rate (as long as handling a single event never takes >16ms) but the UI will still lag, because your input hasn't been handled yet, for instance.

Because I don't think there is a reasonable way to order or prioritise events, I think the only way to both ensure 60fps and not block other events, is just to always be handling events on different threads, and not care if there are unfinished events at the time to render. You could use a thread pool for events to make things a little more efficient, and you could throw an error if a single thread has timed out. I guess using `future`s might help in interacting with the thread pool, but only as syntactic sugar for something that won't even be exposed to the user, I don't think it would really simplify anything.

Maybe the thread pool approach makes sense, but I'm not going to try it or look into it for a while, because the way it works now (main UI thread event handling) is pretty close to universal across UI frameworks, and I think for good reason. The main reason being, as far as I can tell, just avoiding complexity. Being able to modify the UI across multiple threads requires complex synchronization that may or may not be reliable, may impact ergonomics, and will impact performance. I think as a Rust project, there is still potential to eventually tackle this problem since this is all stuff that Rust is uniquely good at, and if we can pull it off it could potentially improve performance, by parallelizing work, but it's a large increase in complexity with a small impact on the API. Because of that I think it should be put off until the API has been iterated over a little more and more important problems have been solved.

For now, as a library user, lag should just be an indicator that you either need to offload work from your event handlers to other threads, or optimize them (or webrender is too slow, in which case you're screwed).

Reply:
@fschutt does that make sense? I'm working on some documentation for this, let me know if there's anything unclear or that I'm missing.

Reply:
Yes, I think it makes sense - multithreading the UI is a big task. So I think you're right. I've made the changes, this PR should be ready to merge.